sympy/physics/units/__init__.py

```diff
@@ -210,35 +210,46 @@
 
 def find_unit(quantity):
     """
-    Return a list of matching units or dimension names.
+    This function returns a list of matching unit and dimension names to a
+    input keyword, dimension or unit. If `quantity` is a string -- keyword/
+    units/dimensions, it returns all instances of units/dimensions with
+    these arguments. This function also returns all units/dimensions
+    which have string input keyword as sub-string. If input `quantity` is
+    a Quantity/Dimension -- unit or dimension, then all units having
+    matching base units or dimensions are returned.
 
-    - If ``quantity`` is a string -- units/dimensions containing the string
-    `quantity`.
-    - If ``quantity`` is a unit or dimension -- units having matching base
-    units or dimensions.
+    Parameters
+    ==========
+
+    quantity : string, Dimension or Quantity
 
     Examples
     ========
 
     >>> from sympy.physics import units as u
     >>> u.find_unit('charge')
-    ['C', 'coulomb', 'coulombs', 'planck_charge']
+    ['C', 'charge', 'coulomb', 'coulombs', 'planck_charge']
     >>> u.find_unit(u.charge)
     ['C', 'coulomb', 'coulombs', 'planck_charge']
     >>> u.find_unit("ampere")
-    ['ampere', 'amperes']
+    ['A', 'ampere', 'amperes', 'planck_current']
     >>> u.find_unit('volt')
-    ['volt', 'volts', 'electronvolt', 'electronvolts', 'planck_voltage']
+    ['V', 'v', 'volt', 'volts', 'voltage', 'electronvolt', 'electronvolts', 'planck_voltage']
+    >>> u.find_unit('energy')
+    ['J', 'eV', 'joule', 'energy', 'joules', 'electronvolt', 'electronvolts', 'planck_energy', 'planck_energy_density']
     >>> u.find_unit(u.inch**3)[:5]
     ['l', 'cl', 'dl', 'ml', 'liter']
+
     """
     import sympy.physics.units as u
     rv = []
     if isinstance(quantity, string_types):
-        rv = [i for i in dir(u) if quantity in i and isinstance(getattr(u, i), Quantity)]
         dim = getattr(u, quantity)
-        if isinstance(dim, Dimension):
-            rv.extend(find_unit(dim))
+        if isinstance(dim, Quantity): dim = dim.dimension
+        for i in dir(u):
+            quant_attr = getattr(u, i)
+            if quantity in i or (isinstance(quant_attr, Quantity) and quant_attr.dimension == dim):
```

Comment:
I don't think it should return `Dimension`s.

Reply:
True, as name of function `find_unit` itself implies to find all instances of that unit. However, original `docstring` states that function should return all `Dimensions`.

Reply:
A separate function like `find_dimension` can be added to return all the dimensions. Please suggest.

Reply:
However, the results of these functions (`find_unit` and `find_dimension`) would be quite alike, which may make this exercise quite dispensable.